File: Rod Cutting.java
class Solution{
    public int cutRod(int price[], int n) {
        //code here
        int[] np=new int[price.length+1];
        for(int i=0;i<price.length;i++){
            np[i+1]=price[i];
        }
        int[] dp=new int[np.length];
        dp[0]=0;
        dp[1]=np[1];
        for(int i=2;i<dp.length;i++){
            //left right strategy
            dp[i]=np[i];
            int li=1;
            int ri=i-1;
            while(li<=ri){
                if(dp[li]+dp[ri] > dp[i]){
                    dp[i]=dp[li]+dp[ri];
                }
                li++;
                ri--;
            }
        }
        return dp[dp.length-1];
    }
}

File: Rod Cutting.cpp
class Solution{
    public:
    int solve(int n ,int need,int sum,int price[],vector<int>&dp){
        if(need==0) return sum;
        if(dp[need-1]!= -1) return sum+dp[need-1];
        int maxi=0;
        for(int j=1;j<=need;j++){
            maxi= max(maxi,solve(n,need-j,sum+price[j-1],price,dp));
        }
        dp[need-1]=maxi-sum;
        return maxi;
    }
    int cutRod(int price[], int n) {
        vector<int>dp(n,-1);
        int x =solve(n,n,0,price,dp);
        return x;
    }
};
